checks.d/agent_metrics.py

```diff
@@ -0,0 +1,150 @@
+import os
+import threading
+
+# 3p
+try:
+    import psutil
+except ImportError:
+    psutil = None
+
+# project
+from checks import AgentCheck
+from config import _is_affirmative
+
+MAX_THREADS_COUNT = 50
+MAX_COLLECTION_TIME = 30
+MAX_EMIT_TIME = 5
+MAX_CPU_PCT = 10
+
+
+class UnsupportedMetricType(Exception):
+    """
+    Raised by :class:`AgentMetrics` when a metric type outside outside of AgentMetrics.ALLOWED_METRIC_TYPES
+    is requested for measurement of a particular statistic
+    """
+    def __init__(self, metric_name, metric_type):
+        message = 'Unsupported Metric Type for {0} : {1}'.format(metric_name, metric_type)
+        Exception.__init__(self, message)
+
+class AgentMetrics(AgentCheck):
+    """
+    New-style version of `CollectorMetrics`
+    Gets information about agent performance on every collector loop
+    """
+
+    def __init__(self, *args, **kwargs):
+        AgentCheck.__init__(self, *args, **kwargs)
+        self._collector_payload = {}
+        self._metric_context = {}
+
+    def _psutil_config_to_stats(self, instance):
+        """
+        Reads `init_config` for `psutil` methods to call on the current process
+        Calls those methods and stores the raw output
+
+        :returns a dictionary of statistic_name: value
+        """
+        process_metrics = instance.get('process_metrics', self.init_config.get('process_metrics', None))
+        if not process_metrics:
+            self.log.error('No metrics configured for AgentMetrics check!')
+            return {}
+
+        methods, metric_types = zip(
+            *[(p['name'], p.get('type', 'gauge'))
```

Comment:
Nitpick but you should use constants for the metric types.

They are actually already existing:
https://github.com/DataDog/dd-agent/blob/master/checks/metric_types.py